src/libraries/System.Text.Json/Common/JsonSeparatorNamingPolicy.cs

```diff
@@ -21,149 +28,154 @@ public sealed override string ConvertName(string name)
+                        }
+                        break;
 
-                    continue;
-                }
+                    default:
+                        // Non-alphanumeric characters (including the separator character itself)
+                        // are written as-is to the output and reset the separator state.
+                        // E.g. 'ABC???def' maps to 'abc???def' in snake_case.
 
-                if (index + 1 < chars.Length)
-                {
-                    char next = chars[index + 1];
-                    CharCategory currentCategory = currentCategoryUnicode switch
-                    {
-                        UnicodeCategory.LowercaseLetter => CharCategory.Lowercase,
-                        UnicodeCategory.UppercaseLetter => CharCategory.Uppercase,
-                        _ => previousCategory
-                    };
-
-                    if (currentCategory == CharCategory.Lowercase && char.IsUpper(next) ||
-                        next == '_')
-                    {
-                        WriteWord(chars.Slice(first, index - first + 1), ref result);
-
-                        previousCategory = CharCategory.Boundary;
-                        first = index + 1;
-
-                        continue;
-                    }
-
-                    if (previousCategory == CharCategory.Uppercase &&
-                        currentCategoryUnicode == UnicodeCategory.UppercaseLetter &&
-                        char.IsLower(next))
-                    {
-                        WriteWord(chars.Slice(first, index - first), ref result);
-
-                        previousCategory = CharCategory.Boundary;
-                        first = index;
-
-                        continue;
-                    }
-
-                    previousCategory = currentCategory;
+                        WriteChar(current, ref destination);
+                        state = SeparatorState.NotStarted;
+                        break;
                 }
             }
 
-            WriteWord(chars.Slice(first), ref result);
-
-            name = result.Slice(0, resultUsedLength).ToString();
+            name = destination.Slice(0, charsWritten).ToString();
 
             if (rentedBuffer is not null)
             {
-                result.Slice(0, resultUsedLength).Clear();
+                destination.Slice(0, charsWritten).Clear();
                 ArrayPool<char>.Shared.Return(rentedBuffer);
             }
 
             return name;
+
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            void WriteChar(char value, ref Span<char> destination)
+            {
+                if (charsWritten == destination.Length)
+                {
+                    ExpandBuffer(ref destination);
+                }
+
+                destination[charsWritten++] = value;
+            }
+
+            void ExpandBuffer(ref Span<char> destination)
+            {
+                int newSize = checked(destination.Length * 2);
```

Comment:
>int newSize = checked(destination.Length * 2); [](http://example.com/codeflow?start=16&length=46)

do we need to check the newSize if overflow occurred? or we expect ArrayPool<char>.Shared.Rent will throw when this happen?

Reply:
`checked` prevents overflows. The runtime will do it for you and throw an exception.

Reply:
I would guess passing a negative number to `ArrayPool.Rent` would throw some kind of ArgumentException, so it's preferable if we catch it here and throw OverflowException.